### DISCUSSION TITLE
v2.1 how to use DB migration?

### DISCUSSION THREAD

**Discussion by Blue-Nitro** [oqtane/oqtane.framework#1456] [Q&A] [answered]
How to use DB migration on modules when I have add more fields in data model and want to update database

**Comment by sbwalker** [comment, marked as answer]
Modules can use migrations the same way that the core framework utilizes them. You can see plenty of examples of different types of migrations in this folder: https://github.com/oqtane/oqtane.framework/tree/dev/Oqtane.Server/Migrations/Tenant. If you are interested in seeing an example of a module which uses migrations you can generate one dynamically using the Module Creator or you can look at the Blog module which was upgraded for 2.1: https://github.com/oqtane/oqtane.blogs

**Reply by Blue-Nitro** [reply]
Nice I got it, thank you. And how to rename DB table on new version of module?

**Reply by sbwalker** [reply]
@Blue-Nitro Oqtane 2.1 uses EF Core - so it should be possible to perform any code-first migration - just search for "EF Core table rename"  in your favorite search engine

**Reply by ionsource** [reply]
I see the many migration examples, but it's not clear how the migrations were generated. Typically I use "Add-Migration" command to automatically create migration scripts, but using this command on an Oqtane module resulted in:
- Build Started
- Build succeeded  
- Unable to create an object of type 'TestContext'. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728

How were the example migration scripts created?

**Reply by sbwalker** [reply]
Due to the multi-tenant and multi-database nature of Oqtane you cannot use the command line migration tool as it only handles simple scenarios ( ie. single tenant, single database ). The core migrations were created by hand - and if you review their contents you will see that they are not complicated.

**Reply by ionsource** [reply]
Ok, makes sense because of the complexity of multi-tenant and multi-database support. Just wanted to make sure before creating. Thanks!
